File: leetcode/python/1143_longest_common_subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m = len(text1)
        n = len(text2)
        dp = [[0] * (n+1) for _ in range(m+1)]
        for col in range(n-1, -1, -1):
            for row in range(m-1, -1, -1):
                if text2[col] == text1[row]:
                    dp[row][col] = 1 + dp[row+1][col+1]
                else:
                    dp[row][col] = max(dp[row+1][col], dp[row][col+1])
        return dp[0][0]

s = Solution()
print(s.longestCommonSubsequence(text1 = "abcde", text2 = "ace"))
print(s.longestCommonSubsequence(text1 = "abc", text2 = "abc"))
print(s.longestCommonSubsequence(text1 = "abc", text2 = "def"))

File: leetcode/rust/src/1143_longest_common_subsequence.rs
impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let mut dp = vec![vec![0; text2.len() + 1]; text1.len() + 1];
        
        let text1 = text1.as_bytes();
        let text2 = text2.as_bytes();
        for i in 1..=text1.len() {
            for j in 1..=text2.len() {
                if text1[i - 1] == text2[j - 1] {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                }
                else {
                    dp[i][j] = dp[i - 1][j].max(dp[i][j - 1]);
                }
            }
        }
        dp[text1.len()][text2.len()]
    }
}